### ISSUE TITLE
Payment Gateways that are compatible with Blocks show as incompatible

### ISSUE THREAD

**Issue by paymentplugins** [woothemes/woocommerce#44712]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

Payment Gateways, such as Payment Plugins for PayPal WooCommerce, which are compatible with the cart and checkout block show as incompatible on the Checkout or Cart block edit page. Gateways coded by WooCommerce don't have this compatibility notice displayed which is very interesting.

<img width="1505" alt="image" src="https://github.com/woocommerce/woocommerce/assets/17156041/fa304965-3bb1-445f-8f0b-9e8accc9a5a2">

### Expected behavior

If a plugin registers compatibility via the `\Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility` method, then no notices should be shown to users that it's incompatible. What's interesting is the `wcSetttings.incompatibleExtensions` array is empty, which means no notice should be shown. That would indicate this notice is coming from somewhere else.

<img width="636" alt="image" src="https://github.com/woocommerce/woocommerce/assets/17156041/40c6447c-0e64-40f9-922a-1f6c05db0072">

### Actual behavior

Despite the plugin correctly registering compatibility with `cart_checkout_blocks`, WooCommerce still displays a notice to users that the plugin is incompatible. The client side variable `wcSetttings.incompatibleExtensions` is empty and yet the notice still appears.

Here you can see the PayPal plugin is listed as compatible:
<img width="962" alt="image" src="https://github.com/woocommerce/woocommerce/assets/17156041/ad22a342-7966-43cf-9d98-e1dbe3999a37">

The `$incompatible_extensions` variable is empty:
<img width="846" alt="image" src="https://github.com/woocommerce/woocommerce/assets/17156041/b07c5880-dca4-4693-a78c-2af322194c07">

### Steps to reproduce

1. Download Payment Plugins for PayPal WooCommerce
2. Activate plugin and navigate to the WordPress > Pages > Edit Page for the Checkout Block
3. Observe the notice stating the plugin is incompatible with the Checkout Block

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): http://localhost:8888/paymentplugins
Site address (URL): http://localhost:8888/paymentplugins
WC Version: 8.6.0
REST API Version: ✔ 8.6.0
WC Blocks Version: ✔ 11.8.0-dev
Action Scheduler Version: ✔ 3.7.1
Log Directory Writable: ✔
WP Version: 6.4.3
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.46 (Unix) OpenSSL/1.0.2u PHP/8.0.8 mod_wsgi/3.5 Python/2.7.18 mod_fastcgi/mod_fastcgi-SNAP-0910052141 mod_perl/2.0.11 Perl/v5.30.1
PHP Version: 8.0.8
PHP Post Max Size: 20 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.68.0
OpenSSL/1.0.2u

SUHOSIN Installed: –
MySQL Version: 5.7.34
Max Upload Size: 20 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.5.2
WC Database Prefix: wp_
Total Database Size: 8558.35MB
Database Data Size: 5040.13MB
Database Index Size: 3518.22MB
wp_woocommerce_sessions: Data: 0.03MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.38MB + Index: 0.16MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 3.52MB + Index: 3.94MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 3249.53MB + Index: 2326.61MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 1761.53MB + Index: 1173.02MB + Engine InnoDB
wp_brizy_logs: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_bwf_ab_experiments: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_bwf_actions: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_bwf_action_claim: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_bwf_contact: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_bwf_contact_meta: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_bwf_conversion_tracking: Data: 0.02MB + Index: 0.20MB + Engine InnoDB
wp_bwf_funnelmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_bwf_funnels: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_bwf_optin_entries: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_bwf_wc_customers: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_cartflows_visits: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_cartflows_visits_meta: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_cfw_acr_carts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 1.52MB + Index: 0.84MB + Engine InnoDB
wp_e_events: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_e_notes: Data: 0.02MB + Index: 0.17MB + Engine InnoDB
wp_e_notes_users_relations: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_e_submissions: Data: 0.02MB + Index: 0.27MB + Engine InnoDB
wp_e_submissions_actions_log: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_e_submissions_values: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 7.02MB + Index: 0.20MB + Engine InnoDB
wp_postmeta: Data: 8.47MB + Index: 5.89MB + Engine InnoDB
wp_posts: Data: 2.45MB + Index: 0.39MB + Engine InnoDB
wp_sheerid_programs: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_sheerid_verifications: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_snippets: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.09MB + Index: 0.09MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.09MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.08MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_od_time_framemeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_od_time_frames: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_orders: Data: 0.44MB + Index: 0.58MB + Engine InnoDB
wp_wc_orders_meta: Data: 2.41MB + Index: 2.84MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.30MB + Index: 0.38MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.20MB + Index: 0.14MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.09MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.06MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wdr_order_discounts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wdr_order_item_discounts: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wdr_rules: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wfacp_stats: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wfco_report_views: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wfob_stats: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wfocu_event: Data: 0.06MB + Index: 0.05MB + Engine InnoDB
wp_wfocu_event_meta: Data: 0.16MB + Index: 0.00MB + Engine InnoDB
wp_wfocu_session: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_gzd_dhl_im_products: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_gzd_dhl_im_product_services: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_gzd_packaging: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_gzd_packagingmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_gzd_shipmentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_gzd_shipments: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_woocommerce_gzd_shipment_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_gzd_shipment_items: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_gzd_shipment_labelmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_gzd_shipment_labels: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_gzd_shipping_provider: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_gzd_shipping_providermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woodmart_wishlists: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woodmart_wishlist_products: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wpforms_payments: Data: 0.02MB + Index: 0.14MB + Engine InnoDB
wp_wpforms_payment_meta: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wpforms_tasks_meta: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpmailsmtp_debug_events: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpmailsmtp_tasks_meta: Data: 0.02MB + Index: 0.00MB + Engine InnoDB

### Post Type Counts ###

attachment: 14
bfwc_subscription: 29
brizy-project: 1
cartflows_flow: 1
cartflows_step: 2
e-landing-page: 1
elementor_library: 6
page: 15
post: 1
product: 9
product_variation: 3
revision: 169
shop_coupon: 5
shop_order: 1240
shop_order_placehold: 783
shop_order_refund: 63
shop_subscription: 112
wfacp_checkout: 1
wffn_ty: 1
wfocu_funnel: 3
wfocu_offer: 3
wp_global_styles: 3
wp_navigation: 1
wp_template: 2

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (7) ###

Code Snippets: by Code Snippets Pro – 3.6.2
Elementor Pro: by Elementor.com – 3.18.1
Elementor: by Elementor.com – 3.19.2
Heartbeat Control by WP Rocket: by WP Rocket – 2.0.1
Payment Plugins for PayPal WooCommerce: by Payment Plugins
support@paymentplugins.com – 1.0.44
WooCommerce: by Automattic – 8.6.0

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: braintree-subscription (braintree-subscription)
external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to Woo.com: –
Enforce Approved Product Download Directories: –
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #16 - /checkout-block/
My account: #9 - /my-account/
Terms and conditions: ❌ Page visibility should be public

### Theme ###

Name: Storefront
Version: 4.5.4
Author URL: https://woo.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Subscriptions ###

WCS_DEBUG: ✔ No
Subscriptions Mode: ✔ Live
Subscriptions Live URL: http://localhost:8888/paymentplugins
Subscriptions-core Library Version: 5.7.1
Subscription Statuses: wc-active: 13
wc-cancelled: 3
wc-on-hold: 45

WooCommerce Account Connected: ❌ No
Report Cache Enabled: ✔ Yes
Cache Update Failures: ✔ 0 failure

### Store Setup ###

Country / State: United States (US) — California

### Subscriptions by Payment Gateway ###

other: wc-on-hold: 2
PayPal Gateway By Payment Plugins: wc-cancelled: 3
wc-on-hold: 27

### Payment Gateway Support ###

PayPal Gateway By Payment Plugins: tokenization
products
subscriptions
add_payment_method
subscription_cancellation
multiple_subscriptions
subscription_amount_changes
subscription_date_changes
default_credit_card_form
refunds
pre-orders
subscription_payment_method_change_admin
subscription_reactivation
subscription_suspension
subscription_payment_method_change_customer

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
product-virtual-downloadable
product-external-affiliate
product-grouped
product-linked
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
product-pre-publish-modal
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2024-02-17 05:22:54 +00:00
Options: ✔
Notes: 151
Onboarding: skipped

### Action Scheduler ###

Complete: 8,492,768
Oldest: 2023-04-04 04:08:23 +0000
Newest: 2024-02-16 06:14:44 +0000

Failed: 249
Oldest: 2023-02-17 21:27:35 +0000
Newest: 2024-02-16 02:59:50 +0000

Pending: 14
Oldest: 2024-02-16 21:58:19 +0000
Newest: 2024-03-12 04:55:09 +0000

### Status report information ###

Generated at: 2024-02-16 17:42:15 +00:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by wavvves** [comment]
Hi @paymentplugins, thank you for reaching out.

I had a [peak at your svn repo](https://plugins.svn.wordpress.org/pymntpl-paypal-woocommerce/trunk/src/Main.php), and have you tried inserting the class existence check within the hook?

Example

```php
private function declare_features() {
		add_action( 'before_woocommerce_init', function () {
			if ( class_exists( '\Automattic\WooCommerce\Utilities\FeaturesUtil' ) ) {
				try {
					\Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', $this->file, true );
					\Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'cart_checkout_blocks', $this->file, true );
				} catch ( \Exception $e ) {
				}
			}
		} );
	}
```

Just to make sure it is not skipping adding the action just because the class is not available yet.

**Comment by senadir** [comment]
Hey @paymentplugins I found the issue in your plugin.

To decide what to show on the notice, we do the following:

- Fetch all plugins that declare themselves incompatible (via `incompatibleExtensions` param).
- Compare all enabled payment methods, and compare them with registered payment methods in Checkout block (we diff the result of this call https://github.com/woocommerce/woocommerce//blob/0c79d1b6a780d0270cfde3d32cc43ebe5fa7d3ac/plugins/woocommerce/src/Blocks/BlockTypes/Checkout.php#L383-L391 and `wp.data.select('wc/store/payment').getAvailablePaymentMethods()` and `wp.data.select('wc/store/payment').getAvailableExpressPaymentMethods()`.

In your case, it seems, server side, you're registering PayPal with `ppcp` ID:

<img width="550" alt="image" src="https://github.com/woocommerce/woocommerce/assets/6165348/9bd38bbd-2161-4658-9555-4e09994346e4">


But client-side (Checkout block), we only see `paymentplugins_ppcp_express`, but not `ppcp`, which it seems is behind some flag.

So, to recap, make sure your payment method is loading fine in the client in Gutenberg for it to show as registered.

**Comment by rex2630** [comment]
So does this mean that the bug is inside woocommerce or the plugin?
